src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -40,27 +40,35 @@ internal sealed class EventCountersMetrics : EventListener
     /// <param name="options">The options to define the metrics.</param>
     public EventCountersMetrics(EventCountersInstrumentationOptions options)
     {
-        this.options = options;
-
-        while (this.preInitEventSources.TryDequeue(out EventSource eventSource))
+        lock (this.preInitEventSources)
         {
-            if (this.options.ShouldListenToSource(eventSource.Name))
+            this.options = options;
+
+            foreach (EventSource eventSource in this.preInitEventSources)
             {
-                this.EnableEvents(eventSource, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+                if (this.options.ShouldListenToSource(eventSource.Name))
+                {
+                    this.EnableEvents(eventSource);
+                }
             }
+
+            this.preInitEventSources.Clear();
         }
     }
 
     /// <inheritdoc />
-    protected override void OnEventSourceCreated(EventSource source)
+    protected override void OnEventSourceCreated(EventSource eventSource)
     {
-        if (this.options == null)
+        lock (this.preInitEventSources)
```

Comment:
Use double-checked locking and remove the else condition from inside the lock. Something like this:

```csharp
        if (this.options == null)
        {
            lock (this.preInitEventSources)
            {
                if (this.options == null)
                {
                    this.preInitEventSources.Enqueue(source);
                }
            }
        }
        else if (this.options.ShouldListenToSource(source.Name))
        {
            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
        }
```

Reply:
Why?
That would just reintroduc the race condition:
- Thread A sees `this.options == null`
- Thread B runs the constructor and sets `options`
- Thread A checks `this.options == null` again but that now fails as the options are set
- The `EventSource` thread A just processed gets dropped silently

Reply:
> Why?

This avoids the thread having to acquire a lock on every `OnEventSourceCreated` call. We don't need the threads to acquire a lock once the options are set.


> Thread A checks this.options == null again but that now fails as the options are set

Umm yeah you could add the else condition inside the lock statement as well to fix that.

Reply:
> This avoids the thread having to acquire a lock on every OnEventSourceCreated call.

I don't see why that's a problem though, this method is only called a dozen times in a given process. Complicating the locking algorithm is not worth it in this case, uncontended lock acquisitions are extremely cheap anyway.

Reply:
The change doesn't seem that complicated to me. Having said that, I don't have a strong opinion on this. I'm okay with always locking in this case.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Internal/SelfDiagnosticsEventListener.cs similar pattern being followed already